### PR TITLE
Add collision strategy

### LINKED ISSUE #11990: feature request: Import API: add `overwrite` to `collision_strategy`
In cartoframes development, we rely on the Import API for updating a table in a user's account. 

Having a `collision_strategy=overwrite` option would make our development cleaner and allow us to offload the complexity of data/map management to CARTO's APIs instead of having to rely on the unintended consequences of dropping a table that a visualization is attached to.

- [ ] Strategy
  - [ ] If the schema does not exactly match, raise an error.
- [ ] Documentation

cc @juanignaciosl

### PR DESCRIPTION
**Pull request by amartinj**
Closes #11990

### CODE REVIEW

File: app/models/data_import.rb [hunk 1]
```diff
@@ -699,6 +699,7 @@ def new_importer
                                                        overviews_creator,
                                                        destination_schema, public_user_roles)
     end
+    importer.overwrite_table = collision_strategy == Carto::DataImportConstants::COLLISION_STRATEGY_OVERWRITE
```
Comment (juanignaciosl) [inline app/models/data_import.rb]: Minor: I'd rename this to `overwrite_on_name_collision`, because you don't actually know if the table will be renamed. Feel free to ignore, though.

It'd be really nice to have it at the constructor instead of having an accessor. The constructor receives a hash. Have you considered refactoring the constructor to named arguments? It should be straightforward, given that it already receives a hash? Not 100% sure about this suggestion, just give it a try.

File: app/models/table_registrar.rb [hunk 2]
```diff
@@ -7,14 +7,18 @@ def initialize(user, table_klass=nil)
       @table_klass  = table_klass
     end
 
-    def register(table_name, data_import_id)
-      self.table                    = table_klass.new
-      table.user_id                 = user.id
+    def register(table_name, data_import_id, skip_table_creation = false)
```
Comment (juanignaciosl) [inline app/models/table_registrar.rb]: Use a named parameter instead of a positional one. No need to refactor the two first parameters, though.

File: app/connectors/importer.rb [hunk 3]
```diff
@@ -70,17 +74,33 @@ def register(result)
         # Sanitizing table name if it corresponds with a PostgreSQL reseved word
         result.name = Carto::DB::Sanitize.sanitize_identifier(result.name)
 
-        runner.log.append("Before renaming from #{result.table_name} to #{result.name}")
+        log("Before renaming from #{result.table_name} to #{result.name}")
+
+        name = Carto::ValidTableNameProposer.new.propose_valid_table_name(result.name, taken_names: [])
```
Comment (juanignaciosl) [inline app/connectors/importer.rb]: As `Carto::ValidTableNameProposer.new.propose_valid_table_name` already calls `Carto::DB::Sanitize.sanitize_identifier`, can't we remove the call at line 75?

File: app/connectors/importer.rb [hunk 4]
```diff
@@ -70,17 +74,33 @@ def register(result)
...
-        runner.log.append("Before moving schema '#{name}' from #{ORIGIN_SCHEMA} to #{@destination_schema}")
-        move_to_schema(result, name, ORIGIN_SCHEMA, @destination_schema)
+        database.transaction do
+          begin
+            if overwrite
+              log("Dropping destination table: #{name}")
+              drop("#{@destination_schema}.#{name}")
+              log("Removing metadata for table #{name}")
+            end
+            result.name = name
```
Comment (juanignaciosl) [inline app/connectors/importer.rb]: Minor: as keeping transactions as small as possible is nice, and in order to be more consistent with previous behaviour, keeping `result.name = name` outside the transaction might be better.

File: spec/models/data_import_spec.rb [hunk 5]
```diff
@@ -61,6 +61,48 @@
     data_import.error_code.should == 2011
   end
 
+  it 'should overwrite dataset if collision_strategy is set accordingly' do
```
Comment (juanignaciosl) [inline spec/models/data_import_spec.rb]: Nice test, but checking table content would be interesting. Could you change the data between calls and add a check? For example, increasing 1 row between calls, and checking that the existing tables and the imported one (if any) have the right amount of rows.

File: app/models/table_registrar.rb [hunk 6]
```diff
@@ -7,14 +7,18 @@ def initialize(user, table_klass=nil)
...
+      self.table = table_klass.new(
+        user_table: skip_table_creation ? Carto::UserTable.where(user_id: @user.id, name: table_name).first : nil
+      )
+      table.user_id = user.id unless skip_table_creation
       # INFO: we're not creating but registering an existent table, so we want fixed, known name
       table.instance_eval { self[:name] = table_name }
       table.migrate_existing_table  = table_name
       table.data_import_id  = data_import_id
       set_metadata_from_data_import_id(table, data_import_id)
+      table.update_cdb_tablemetadata if skip_table_creation
```
Comment (juanignaciosl) [inline app/models/table_registrar.rb]: Minor: consider joining both lines inside an `if skip_table_creation` conditional.

File: app/models/table_registrar.rb [hunk 7]
```diff
@@ -7,14 +7,18 @@ def initialize(user, table_klass=nil)
...
+      table.cartodbfy if skip_table_creation
```
Comment (juanignaciosl) [inline app/models/table_registrar.rb]: I'm pretty sure that we talked about this offline, but I can't remember it. How come a cartodbfication is needed? The normal import flow should've already taken care of this.

I get the need for `update_cdb_tablemetadata`. Nevertheless, I'm not sure if TableRegistrar is the place for this. Maybe it should be done outside, specially if we could remove the need for TableRegistrar changes (I don't know if it's possible).

File: app/models/table_registrar.rb [hunk 8]
```diff
@@ -7,14 +7,18 @@ def initialize(user, table_klass=nil)
       @table_klass  = table_klass
     end
 
-    def register(table_name, data_import_id)
-      self.table                    = table_klass.new
-      table.user_id                 = user.id
+    def register(table_name, data_import_id, skip_table_creation = false)
+      self.table = table_klass.new(
+        user_table: skip_table_creation ? Carto::UserTable.where(user_id: @user.id, name: table_name).first : nil
```
Comment (juanignaciosl) [inline app/models/table_registrar.rb]: What happens if we pass `nil`, or if we _always_ pass the `UserTable`? I don't know if this is possible, but I'd like to push this so we could avoid changes at TableRegistrar, because from a logic point of view, the normal and the override scenario _might_ be supported just alike.

**Comment by javitonino** [comment]
Do we want to treat this more as a sync, or more as an import?

Some things that the [sync code](https://github.com/CartoDB/cartodb/blob/master/app/models/synchronization/adapter.rb#L31-L40) does:
- Keep indices
- Keep permissions (we probably want this, at least)
- Keep overviews

Something to consider

**Comment by juanignaciosl** [comment]
Probably all of them. Closer to syncs not for the sake of being more like a sync, but because of that making sense (or being needed!).

File: app/models/data_import.rb (hunk 1 above)
Comment (amartinj) [reply app/models/data_import.rb]: You do know because both strategy is set to `overwrite` and the name is already taken

File: app/connectors/importer.rb (hunk 3 above)
Comment (amartinj) [reply app/connectors/importer.rb]: It's a different purpose. One is called as if no names were already taken and the second one is called with the taken names (or not), depending on the collision strategy

Comment (juanignaciosl) [reply app/connectors/importer.rb]: Ok, I'll have a second thought in the next CR.

File: spec/models/data_import_spec.rb [hunk 9]
```diff
@@ -61,6 +61,51 @@
...
+    data_import.state.should eq 'complete'
+    data_import.table_name.should eq 'walmart_latlon'
+
+    data_import = create_import(false, false)
+    data_import.run_import!
+    carto_user.reload
+    carto_user.visualizations.count.should eq 3
+    data_import.state.should eq 'complete'
+    data_import.table_name.should eq 'walmart_latlon_1'
+    byebug
```
[the remark this answers is not in the document]
Comment (javitonino) [reply spec/models/data_import_spec.rb]: I killed the affected CI run, because the worker got stuck in the byebug menu xD

**Comment by amartinj** [comment]
retest this, please

File: app/connectors/importer.rb [hunk 10]
```diff
@@ -22,9 +27,10 @@ class Importer
       # @param data_import_id String UUID
       # @param destination_schema String|nil
       # @param public_user_roles Array|nil
-      def initialize(runner, table_registrar, quota_checker, database, data_import_id,
-                     overviews_creator,
-                     destination_schema = DESTINATION_SCHEMA, public_user_roles=[CartoDB::PUBLIC_DB_USER])
+      def initialize(runner: nil, table_registrar: nil, quota_checker: nil, database: nil, data_import_id: nil,
+                     overviews_creator: nil,
+                     destination_schema: DESTINATION_SCHEMA, public_user_roles: [CartoDB::PUBLIC_DB_USER],
+                     collision_strategy: Carto::DataImportConstants::COLLISION_STRATEGY_SKIP)
```
Comment (juanignaciosl) [inline app/connectors/importer.rb]: Setting `collision_strategy` to `COLLISION_STRATEGY_SKIP` changes the default behavior. By default, the strategy is just `nil`, which is "search for an available name appending _X". There's arguably a refactor making this explicit, but unless I'm missing something, that shouldn't be set to `COLLISION_STRATEGY_SKIP`. 

As a rule of thumb, take care with behavior changes.

File: app/connectors/importer.rb [hunk 11]
```diff
@@ -70,17 +77,46 @@ def register(result)
         # Sanitizing table name if it corresponds with a PostgreSQL reseved word
         result.name = Carto::DB::Sanitize.sanitize_identifier(result.name)
 
-        runner.log.append("Before renaming from #{result.table_name} to #{result.name}")
-        name = rename(result, result.table_name, result.name)
-        result.name = name
+        log("Before renaming from #{result.table_name} to #{result.name}")
+
+        name = Carto::ValidTableNameProposer.new.propose_valid_table_name(result.name, taken_names: [])
```
Comment (juanignaciosl) [inline app/connectors/importer.rb]: I already asked this: 

> As Carto::ValidTableNameProposer.new.propose_valid_table_name already calls Carto::DB::Sanitize.sanitize_identifier, can't we remove the call at line 75?

you answered this:

> It's a different purpose. One is called as if no names were already taken and the second one is called with the taken names (or not), depending on the collision strategy

I'm sorry about going on about this again, but I'm not sure. I'll try to explain my point:

`register` currently runs (1) `sanitize_identifier`on `name`, and then runs (2) `propose_valid_table_name`. This method calls (2.1) `sanitize_identifier` as well:

https://github.com/CartoDB/cartodb/blob/3c577ce073c951f75c45fad8d4c963abd19f30e4/lib/carto/valid_table_name_proposer.rb#L13-L16

So, unless I'm wrong, the result of this is calling `sanitize_identifier` twice, and then resume. 

I don't get your point. You said that "One is called as if no names were already taken and the second one is called with the taken names (or not), depending on the collision strategy", but that's not what I see: `sanitize_identifier` is always* called twice, no matter taken names or strategy.

_*: there's one scenario where it's not called twice. If you pass `nil` you'd get `DEFAULT_TABLE_NAME` because of the check at `propose_valid_table_name`, but I think that it's an exception.

I think that the call to `sanitize_identifier` at `register` can be removed, because `propose_valid_table_name` already does that, and it's a nice assumption about whose concern it is (it's logic that `propose_valid_table_name` should sanitize).

What do you think? I might be missing something (again), but I couldn't just skip the discussion :_).

PS: maybe you thought that I was asking about the double call to `propose_valid_table_name` instead of `sanitize_identifier`?

File: app/connectors/importer.rb [hunk 12]
```diff
@@ -70,17 +77,46 @@ def register(result)
...
+
+        overwrite = overwrite_table? && taken_names.include?(name)
+        assert_schemas_are_compatible(name) if overwrite
+
+        index_statements = generate_index_statements(@destination_schema, name) if overwrite
+
+        database.transaction do
+          name = rename(result, result.table_name, result.name)
```
Comment (juanignaciosl) [inline app/connectors/importer.rb]: Disclaimer: open question, not sure about this, but it might worth a discussion.

After thinking about the previous question, one thing made me raise my eyebrow. `register` calls `propose_valid_table_name` and then calls `rename`, which calls it again. They don't have the same arguments for `taken_names` in every scenario, but I think that the second call is always "equallly or more restrictive" than the second one, so the first one _might_ not be needed.

You need the result of the first one to do this:

```
        overwrite = overwrite_table? && taken_names.include?(name)
        assert_schemas_are_compatible(name) if overwrite

        index_statements = generate_index_statements(@destination_schema, name) if overwrite
```

Then, `rename` is called, no matter what happens before. Even more, `name` won't change.

So, we run `propose_valid_table_name, run some logic, and then run a totally unrelated `rename`, which in fact can change `name`.

Aaaaaaaaafter all this, the question: wouldn't the result be the same if the first `propose_valid_table_name` be skipped, and the previous three lines moved after `rename`?

I'm not sure about this. It's probably related to your comment about "It's a different purpose. One is called as if no names were already taken and the second one is called with the taken names (or not), depending on the collision strategy" but although I get your point about the intention, I think that, in practice, the first one is not needed.

File: app/connectors/importer.rb (hunk 10 above)
Comment (amartinj) [reply app/connectors/importer.rb]: You are right. Missed it

File: app/connectors/importer.rb (hunk 11 above)
Comment (amartinj) [reply app/connectors/importer.rb]: In fact, I thought you were talking about `propose_valid_table_name`

File: app/connectors/importer.rb (hunk 12 above)
Comment (juanignaciosl) [reply app/connectors/importer.rb]: @amartinj what's your point of view on this?

Comment (amartinj) [reply app/connectors/importer.rb]: Yes, sorry. It was a deep suggestion and I only processed part of it. Currently testing it to see if it breaks anything, but I absolutely agree on your point :)

Comment (amartinj) [reply app/connectors/importer.rb]: Okay, this does not work. The first name proposal is performed just like the first time, and used for evaluating if there is going to be an overwrite, but the second one is called based on this renaming decision. After changing it, a refactor was needed and, despite that, more than 100 test were breaking.

**Comment by javitonino** [comment]
@juanignaciosl I'd prefer if you took a look at this one, so I can be free for acceptance. It is a tricky one, where I feel the clean room approach is important.

File: lib/carto/importer/table_setup.rb [hunk 13]
```diff
@@ -0,0 +1,137 @@
+module Carto
+  module Importer
+    # This module can be used as a mixin as long as the class including it implements the following methods:
```
Comment (juanignaciosl) [inline lib/carto/importer/table_setup.rb]: `user` and `runner` are actually attributes, not methods. Can't we choose _composition over inheritance_ here and create a `TableSetup` class receiving those attributes in the constructor?

File: app/connectors/importer.rb [hunk 14]
```diff
@@ -22,9 +27,10 @@ class Importer
       # @param data_import_id String UUID
       # @param destination_schema String|nil
       # @param public_user_roles Array|nil
-      def initialize(runner, table_registrar, quota_checker, database, data_import_id,
-                     overviews_creator,
-                     destination_schema = DESTINATION_SCHEMA, public_user_roles=[CartoDB::PUBLIC_DB_USER])
+      def initialize(runner: nil, table_registrar: nil, quota_checker: nil, database: nil, data_import_id: nil,
```
Comment (juanignaciosl) [inline app/connectors/importer.rb]: Nice <3

Aren't there mandatory parameters? If there are, move them to the beginning and remove the default value, please.

File: app/connectors/importer.rb [hunk 15]
```diff
@@ -250,10 +276,48 @@ def data_import
...
+        orig_schema = user.in_database.schema(results.first.tables.first, reload:true, schema: ORIGIN_SCHEMA)
+        dest_schema = user.in_database.schema(name, reload:true, schema: user.database_schema)
+        valid = true
+
+        dest_schema.each do |dest_row|
+          next if COLUMNS_NOT_TO_VALIDATE.include?(dest_row[0])
+          break if !valid
+          row_valid = false
+          orig_schema.each do |orig_row|
+            if orig_row[0] == :the_geom && orig_row[0] == dest_row[0] && dest_row[1][:db_type].include?(orig_row[1][:db_type]) ||
```
[the remark this answers is not in the document]
Comment (juanignaciosl) [reply app/connectors/importer.rb]: ^ Honor this, please. It's not only long but hard to understand.

It will probably appear as a consequence of making the line shorter, but I'd use parenthesis to make `&&` vs `||` precedent explicit. It's easier to read, once there are many conditions.

File: app/connectors/importer.rb [hunk 16]
```diff
@@ -66,21 +73,48 @@ def run(tracker)
...
+        log("Before renaming from #{result.table_name} to #{result.name}")
 
-        # Sanitizing table name if it corresponds with a PostgreSQL reseved word
-        result.name = Carto::DB::Sanitize.sanitize_identifier(result.name)
+        name = Carto::ValidTableNameProposer.new.propose_valid_table_name(result.name, taken_names: [])
 
-        runner.log.append("Before renaming from #{result.table_name} to #{result.name}")
-        name = rename(result, result.table_name, result.name)
-        result.name = name
+        overwrite = overwrite_table? && taken_names.include?(name)
```
Comment (juanignaciosl) [inline app/connectors/importer.rb]: `overwrite_table?` name is confusing. I'm sorry about being picky with this, but it seems to do more than checking the chosen strategy. I think that `overwrite_strategy?` is simpler and more accurate.

File: app/connectors/importer.rb [hunk 17]
```diff
@@ -250,10 +289,48 @@ def data_import
 
       private
 
+      def assert_schemas_are_compatible(name)
```
Comment (juanignaciosl) [inline app/connectors/importer.rb]: Minor: instead of having a single `name` parameter, receiving two schemas feels more natural.

File: app/connectors/importer.rb [hunk 18]
```diff
@@ -250,10 +289,48 @@ def data_import
...
+            if orig_row[0] == :the_geom && orig_row[0] == dest_row[0] && dest_row[1][:db_type].include?(orig_row[1][:db_type]) ||
+               orig_row[0] == dest_row[0] && orig_row[1][:db_type].convert_to_cartodb_type == dest_row[1][:db_type].convert_to_cartodb_type
+              row_valid = true
+              break
+            end
+          end
+          valid = row_valid
+        end
+
+        raise 'Incompatible schemas' unless valid
```
Comment (juanignaciosl) [inline app/connectors/importer.rb]: Instead of having a method that raises an error, just return true/false and raise unless. Method usage becomes easier to understand, as you don't depend on error raising, and you'll be able to return as soon you get to a invalid check.

File: app/connectors/importer.rb [hunk 19]
```diff
@@ -250,10 +289,48 @@ def data_import
...
+
+      def log(message)
+        runner.log.append(message)
+      end
+
       def exists_user_table_for_user_id(table_name, user_id)
         !Carto::UserTable.where(name: table_name, user_id: user_id).first.nil?
       end
 
+      def taken_names
```
Comment (juanignaciosl) [inline app/connectors/importer.rb]: Moving this from variable to method is a behaviour change.

1.- Is it needed? As it's evaluated twice, it's slower.

2.- Does it work in every situation? As it's evaluated in two different moments, the second one might get different results if tables are created meanwhile. Beyond the happy path, a zip or .carto with several datasets might be a problem. Unlikely, but worths a second thought.

File: app/models/table_registrar.rb [hunk 20]
```diff
@@ -7,9 +7,11 @@ def initialize(user, table_klass=nil)
       @table_klass  = table_klass
     end
 
-    def register(table_name, data_import_id)
-      self.table                    = table_klass.new
-      table.user_id                 = user.id
+    def register(table_name, data_import_id, user_table = nil)
+      self.table = table_klass.new(
+        user_table: user_table
```
Comment (juanignaciosl) [inline app/models/table_registrar.rb]: No need to split this in different lines.

File: spec/models/data_import_spec.rb (hunk 5 above)
Comment (juanignaciosl) [reply spec/models/data_import_spec.rb]: ^ ?

File: app/connectors/importer.rb (hunk 16 above)
Comment (amartinj) [reply app/connectors/importer.rb]: You are right. This evolved. Trees don't let me see the forest

File: app/connectors/importer.rb (hunk 19 above)
Comment (amartinj) [reply app/connectors/importer.rb]: No tables are created between the two calls. Already checked. I changed it because it resulted in a much cleaner code.

The first call only checks for overwrite scenarios while the second one is used for picking a new name for the final table.

I'd leave it as it is since it's way more easy to read

File: app/connectors/importer.rb (hunk 17 above)
Comment (amartinj) [reply app/connectors/importer.rb]: Makes the code more dirty and less encapsulated. Maybe finding a better name...

File: lib/carto/importer/table_setup.rb (hunk 13 above)
Comment (juanignaciosl) [reply lib/carto/importer/table_setup.rb]: @amartinj nice to see that you managed to do this! Please, delete these three lines while I do the full CR. I'd do it myself, but maybe you want to replace them (although I don't think so)

Comment (amartinj) [reply lib/carto/importer/table_setup.rb]: oops, missed that one :/

File: app/connectors/importer.rb [hunk 21]
```diff
@@ -12,6 +13,8 @@ class Importer
       DESTINATION_SCHEMA  = 'public'
       MAX_RENAME_RETRIES  = 20
 
+      COLUMNS_NOT_TO_VALIDATE = [:cartodb_id, :the_geom_webmercator].freeze
```
Comment (juanignaciosl) [inline app/connectors/importer.rb]: This is not obvious for a reader: how come these columns do not need validation? Maybe it's because one of the tables might not be cartodbfied yet, but I'm not sure. A comment here explaining the need would be helpful.

File: app/connectors/importer.rb [hunk 22]
```diff
@@ -11,6 +12,10 @@ class Importer
       ORIGIN_SCHEMA       = 'cdb_importer'
       DESTINATION_SCHEMA  = 'public'
       MAX_RENAME_RETRIES  = 20
+      
+      # The following columns are not validated because we are comparing schemas of a cartodbfied table and one that
+      # is about to be imported.
```
Comment (juanignaciosl) [inline app/connectors/importer.rb]: ❤️  thanks! 🤗

File: app/connectors/importer.rb [hunk 23]
```diff
@@ -66,21 +77,48 @@ def run(tracker)
 
       def register(result)
         @support_tables_helper.reset
+        log("Before renaming from #{result.table_name} to #{result.name}")
```
Comment (juanignaciosl) [inline app/connectors/importer.rb]: I think that the full method body should be within the transaction. There are several db queries (such as taking taken names), so the more that it emcompasses, the better. In addition, you can probably remove the inner `begin .. rescue` block.

File: lib/carto/importer/table_setup.rb [hunk 24]
```diff
@@ -0,0 +1,141 @@
...
+        @user = user
+        @overviews_creator = overviews_creator
+        @log = log
+        @statement_timeout = statement_timeout
+      end
+
+      # Store all indexes to re-create them after "syncing" the table by reimporting and swapping it
+      # INFO: As upon import geom index names are not enforced, they might "not collide" and generate one on the new
+      # import plus the one already existing, so we skip those
+      def generate_index_statements(origin_schema, origin_table_name)
```
Comment (juanignaciosl) [inline lib/carto/importer/table_setup.rb]: Not to do now, but as a reflexion about previous comments and the PR general comment: having a class helps visualizing the cohesion between the methods: most of them are called together. This probably enables us to do a great refactor in the future: making most (all?) methods private and creating a new public one:

```ruby
def setup_table
  # pre-dowhatever-logic
  yield
  # post-whatever-logic
```

This is probably not the best moment to do it (for the CR, being able to check the diff in a line-by-line basis is nice), but maybe next time that we must change somewhere around this refactor could be tackled as a first step.

File: spec/models/data_import_spec.rb [hunk 25]
```diff
@@ -61,6 +61,50 @@
...
+    data_import = create_import(true, true)
+    data_import.run_import!
+    carto_user.reload
+    carto_user.visualizations.count.should eq 3
+    data_import.state.should eq 'complete'
+    data_import.table_name.should eq 'walmart_latlon'
+    data_import.user.in_database["select count(*) from #{data_import.table_name}"].all[0][:count].should eq 2
+  end
+
+  def create_import(overwrite, truncated)
```
Comment (juanignaciosl) [inline spec/models/data_import_spec.rb]: Minor: make these parameters named, please (reading pairs of booleans is not clear).

File: spec/models/data_import_spec.rb [hunk 26]
```diff
@@ -61,6 +61,50 @@
     data_import.error_code.should == 2011
   end
 
+  it 'should overwrite dataset if collision_strategy is set to overwrite' do
```
Comment (juanignaciosl) [inline spec/models/data_import_spec.rb]: This test should test also that, after an overwrite, the content change:

1. Import a file.
2. Check rows.
3. Add a row to the file.
4. Import again, overwriting.
5. Check rows + 1.

Comment (amartinj) [reply spec/models/data_import_spec.rb]: This is already done. Notice these [line](https://github.com/CartoDB/cartodb/pull/12900/files#diff-d315254f793609a1a380530c94a217b1R75) and this [other line](https://github.com/CartoDB/cartodb/pull/12900/files#diff-d315254f793609a1a380530c94a217b1R83) keeping [this](https://github.com/CartoDB/cartodb/pull/12900/files#diff-d315254f793609a1a380530c94a217b1R95) in mind

Comment (juanignaciosl) [reply spec/models/data_import_spec.rb]: Oh, well, I was misled by a small detail. You import the file three times, and the matching override ones are actually the first and the last one. Nevertheless you check the size of the table on the second. It's true that it's _almost_ equivalent, but my OCD would gently ask you to move the `data_import.user.in_database["select count(*) from #{data_import.table_name}"].all[0][:count].should eq 3176` check to the first import (line 73) instead of the second 🤗

File: app/connectors/importer.rb (hunk 23 above)
Comment (amartinj) [reply app/connectors/importer.rb]: some methods, such as index generation cannot be inside the transaction, so moving just some of the post-move operations inside the transaction would violate their order

Comment (juanignaciosl) [reply app/connectors/importer.rb]: Ok! Keeping order is more important, so just forget it. Trying it was interesting, though. Maybe you can add a comment at the beginning of the method so my future self won't suggest this again :-)

**Comment by javitonino** [comment]
🔴 There seems to be something wrong with table registering on overwrite. A failing use case.

1. Import table a
`curl -v -F "file=@a.csv" "https://user.carto.com/api/v1/imports/?api_key=xxx"`
2. Table `a` appears in dashboard
3. Import table a with overwrite
`curl -v -F "collision_strategy=overwrite" -F "file=@a.csv" "https://user.carto.com/api/v1/imports/?api_key=xxx"`
4. Table `a` disappeared in dashboard ⚠️ I see a ghost tables trace: https://rollbar.com/carto/CartoDB/items/4426/occurrences/31946475800/
5. `SELECT CDB_CartodbfyTable('a')`
6. Table `a` reappears in dashboard

**Comment by amartinj** [comment]
@javitonino can you please attach ```a.csv```? Just to make sure I can reproduce it

**Comment by javitonino** [comment]
`echo -e 'a,b\n1,2' > a.csv`

File: app/connectors/importer.rb [hunk 27]
```diff
@@ -66,21 +77,52 @@ def run(tracker)
...
+          index_statements = @table_setup.generate_index_statements(@destination_schema, name)
+        end
 
-        runner.log.append("Before persisting metadata '#{name}' data_import_id: #{data_import_id}")
-        persist_metadata(result, name, data_import_id)
+        database.transaction do
+          rename(result, result.table_name, name)
+          begin
+            if overwrite
+              @table_setup.cartodbfy(name)
```
Comment (javitonino) [inline app/connectors/importer.rb]: You are cartodbfying twice. Here an ...

File: app/connectors/importer.rb [hunk 28]
```diff
@@ -66,21 +77,52 @@ def run(tracker)
...
+          end
+        end
+
+        if overwrite
+          # This is done here instead of doing it inside the transaction because some of the changes would not yet be visible
+          # for creating indexes and fixing oid so, since order is importante here, it has all ben taken out of it.
+          @table_setup.run_index_statements(index_statements, @database)
+          @table_setup.recreate_overviews(name)
+          @table_setup.fix_oid(name)
+          @table_setup.cartodbfy(name)
```
Comment (javitonino) [inline app/connectors/importer.rb]: ... here!

Looking at the synchronization code, it looks like the proper one is the one before overwrite.

Also, it would be nice to have a test to ensure the final table is properly cartodbfied to to avoid this issue happening again (you can look at GhostTableManager code and tests to see how to detect cartodbfied tables).

Comment (amartinj) [reply app/connectors/importer.rb]: hm... 
```
              @table_setup.cartodbfy(name)
              @table_setup.copy_privileges(@destination_schema, name, ORIGIN_SCHEMA, name)
              log("Dropping destination table: #{name}")
              drop("#{@destination_schema}.#{name}")
```

Looks that we are cartodbfying a table that is going to be removed. cartodbfy does not take the schema into account. Maybe we should add an optional parameter schema to be able to define this. The other option would be leaving the cartodbfication for after moving the table...

If we want to stick to the way it's done in the sync, the first one is the way to go.

Thoughts, @javitonino?

Comment (javitonino) [reply app/connectors/importer.rb]: Uhm, the schema thing looks reasonable, but how does the sync work then? It should fail with the same error.

Answering myself, the sync [moves the table into the destination schema](https://github.com/CartoDB/cartodb/pull/12900/files#diff-54660465566f364241ee5767177826c9R39), before cartodbfying it. Maybe we can do the same thing here?

Comment (amartinj) [reply app/connectors/importer.rb]: I was answering myself too when I got your answer. It doesn't allow cartodbfying in the import schema because it needs quotas to be set first:
```
Sequel::DatabaseError: PG::Error: ERROR:  Please set user quota before cartodbfying tables.
CONTEXT:  SQL statement "SELECT cartodb._CDB_check_prerequisites(destschema, reloid)"
PL/pgSQL function cdb_cartodbfytable(text,regclass) line 21 at PERFORM
```
I'll do it the first thing after moving the table from import schema to the user schema

File: spec/models/data_import_spec.rb [hunk 29]
```diff
@@ -135,7 +135,7 @@
   end
 
   def user_tables_should_be_registered
-    Carto::GhostTablesManager.new(@user.id).user_tables_synced_with_db?.should eq true
+    raise 'tables not properly registered' unless Carto::GhostTablesManager.new(@user.id).user_tables_synced_with_db?
```
Comment (javitonino) [inline spec/models/data_import_spec.rb]: I think you can pass a second parameter to should instead:
`Carto::GhostTablesManager.new(@user.id).user_tables_synced_with_db?.should eq(true), "It exploded"`

File: app/connectors/importer.rb [hunk 30]
```diff
@@ -110,10 +109,10 @@ def register(result)
         if overwrite
           # This is done here instead of doing it inside the transaction because some of the changes would not yet be visible
           # for creating indexes and fixing oid so, since order is importante here, it has all ben taken out of it.
+          @table_setup.cartodbfy(name) # This is done here because in the imports it's moved first to the dest schema
```
Comment (javitonino) [inline app/connectors/importer.rb]: You are doing `cartodbfy` after overwrite, which exposes you to a ghost tables race condition:

1. The importer overwrites the table, and, just before this line...
2. Ghost tables run, it will delete the table metadata, since it's not cartodbfied at this point
3. The importer continues and cartodbfies the table
4. Next ghost table run will reregister the table, but the metadata is not recovered (maps, etc.)

`cartodbfy` must be done:
- After moving the new table to the destination schema (or cartodbfy does not work)
- Before dropping the old table (to avoid the race condition described here)
- Before renaming the new table (when it's still named `importer_xxxx`, to avoid another ghost tables race that implies getting two registered tables)

in order to avoid this. So the proper order would be:
1. Move imported table to user schema
2. Cartodbfy
3. Drop the old table
4. Rename the new table.

In other words, in the exact same order than the synchronization code does it.

Also, put `fix_oid` as soon after the transaction as possible, definitely before the indices and overviews. It's important to do it as soon as possible to avoid yet another race condition.

**Comment by javitonino** [comment]
Acceptance is mostly ok. All the major operations work as expected. Tested:
- Syncs (URL/Google drive)
- Import (no strategy, skip strategy, overwrite strategy)
- Schema validation

Everything works, but I'd like to request a last change. The error exposed to the user in case of schema incompatibility is 99999 - Unknown error, which is way too generic. I would like to have a new error code/message for this case.

**Comment by javitonino** [comment]
Working wonders. I'll leave it for tomorrow, not an afternoon deploy.
